docs/rfcs/20210111_centralized_configuration_management.md

```diff
@@ -0,0 +1,588 @@
+- Feature Name: centralized configuration management
+- Status: in-progress
+- Start Date: 2021-01-10
+- Authors:
+- Issue: (one or more # from the issue tracker)
+
+# Executive Summary
+
+Currently we configure redpanda brokers using old-fashioned configuration files
+residing on each node. This approach makes certain day-two operations complicated
+and error prone. One of such examples would be changing the default topic
+replication factor. With current approach in order to change the default
+replication factor one have to edit configuration file on each node and then do
+the cluster rolling restart. It is very important to set the correct (the same)
+value on each of the nodes as a simple typo may lead to undefined behavior.
+
+## What is being proposed
+
+Since redpanda is de-facto a database building block (WAL), we can leverage that
+fact and build simple distributed key value store to handle global cluster
+configuration. Configuration properties that are required only locally and are
+different for each node like f.e. listen_address or node id would be stored in
+an existing `storage::kv_store`.
+
+## Why (short reason)
+
+Storing configuration globally would allow user to simplify cluster management.
+User will be able to change configuration by both the RPK and simple web page
+embedded and served by redpanda. Additionally configuration value will be
+coherent across the cluster nodes. What is also worth mentioning is the support
+for changing some of the configuration parameters in runtime -
+feature that isn't currently supported by redpanda.
+
+## What is the expected outcome?
+
+After the proposed change will be applied we except redpanda to be easily
+configurable with one or all of the following:
+
+- `redpanda` binary command line arguments
+- environment variables
+- `rpk` command
+- REST api exposed by each broker
+
+All global configuration properties should be changeable from each node.
+
+# Reference-level explanation
+## Top level design
+
+Configuration is a map of configuration properties. Configuration interface
+should allow accessing configuration properties from each core. After redpanda
+starts it hydrates configuration properties from different sources.
+Command line arguments always take precedence. After first start redpanda will
```

Comment:
I think it is perfectly fine to require the command line arguments every time. This way it makes it very clear what properties are being used and we don't have any issues of "where is that config set from?"

Reply:
This is very good point, I was thinking of this, in the end we do not have many properties that are node specific. The only properties that would be persisted locally are the `invariant` properties as we have to verify if they changed.